### PR TITLE
refactor: create `fish-future-feature-flags` crate

### PR DESCRIPTION
**Pull request by danielrainer**
Another step in splitting up the main library crate.

At the moment, this breaks tests outside of the new crate, since the
crate changes behavior depending on whether the `test` config is active,
but tests outside the crate get the version with `test` disabled.

I tried to extract this because some other code I want to extract depends on it. However, in the current state it does not work. I haven't looked much into the semantics of the code, but the core problem is that `#[cfg(test)]` does not work across crate boundaries. Maybe we can resolve this by removing the behavioral difference in this case. Otherwise it might be possible to hack around it via some feature setting via `build.rs`, but that does not seem very robust.

Putting it out there in case someone has an idea of how to deal with this problem.

Stacked on #12493, but only to reduce merge conflicts.

### CODE REVIEW

**Comment by krobelus** [comment]
This is an ugly problem (the root cause is that we still have global variables);
for now we could make `FEATURES` use `RefCell`; that should allow us to remove `LOCAL_FEATURES`.
`RefCell` is the next best thing (better than `Mutex`) because it asserts that we only borrow it mutably at specific points in time (in practice at startups).
`scoped_test` can be implemented by storing a copy of `FEATURES` in a local variable, to restore on exit

**Comment by danielrainer** [comment]
I don't think using `RefCell` for `FEATURES` would be sound, since `RefCell` is for single-threaded use cases.

I've pushed an approach which keeps `FEATURES` as is, but uses a thread-local override stack for tests. This allows for extraction, although it comes with a minor performance penalty, because non-test code now has to check the (always empty) override stack before getting the relevant value from `FEATURES`.

I've also explored an alternative approach where `FEATURES` are guarded by `OnceLock`, but that would make handling quite a bit more annoying, especially for tests. It made me realize that we perform feature tests while setting up the `EnvStack` via `src/bin/fish.rs:483`, which eventually reaches `src/common.rs:373`. This might be fine, but it seems a bit unclean to perform feature tests before the features were set. I don't have a solution for it though.

File: Cargo.lock [hunk 1]
```diff
@@ -348,6 +349,13 @@ dependencies = [
  "widestring",
 ]
 
+[[package]]
+name = "fish-feature-flags"
+version = "0.0.0"
+dependencies = [
+ "fish-widestring",
+]
+
```
Comment (krobelus) [inline Cargo.lock:358]: > Prior to this commit, they did this by using ...
 > [..]
 > a new approach to feature overrides in tests is introduced in
 > this commit

I like the commit message conventions used in git.git: use present tense to describe present state ("prior to this commit") and use imperative mood (instead of passive) to describe changes made by this commit. Either is fine.

 > they did this by using thread-local versions of `FEATURES`

(unrelated to this PR) I guess that may not work reliably for tests that spawn new threads, because AIUI, they will not inherit the thread-local override.
Thread-locals has been suggested [here](https://github.com/fish-shell/fish-shell/pull/9881#issuecomment-1627925890) because most tests only use one thread. Hmm.
In future, we might want to change the thread-local override to a global...
If we do, conflicts between tests can still be avoided by calling `let _cleanup = test_init()` which takes a global lock -- so only one of these tests can be running at any given time.
It's a bit weird that test code makes us add thread locals in non-test code paths, but I guess that's okay-ish since the stack is always empty.

 > Instead of having a thread-local version of `FEATURES`, all code,
 > whether test or not, uses the process-wide `FEATURES`.

Ok. This has the potential to uncover latent issues in tests that depend on `FEATURES` but fail to use `test_init()`. Good.

File: crates/feature-flags/src/lib.rs [hunk 2]
```diff
@@ -156,52 +155,34 @@ pub const METADATA: &[FeatureMetadata] = &[
 ];
 
 thread_local!(
-    #[cfg(test)]
-    static LOCAL_FEATURES: RefCell<Option<Features>> = const { RefCell::new(None) };
+    static LOCAL_OVERRIDE_STACK: RefCell<Vec<(FeatureFlag, bool)>> =
```
Comment (krobelus) [inline crates/feature-flags/src/lib.rs:158]: > The former is a fairly minor issue, since I don't think we need nested overrides.

Why make it a Vec instead of Option then?
Should probably only do that if we have a concrete need.

Comment (krobelus) [reply crates/feature-flags/src/lib.rs:158]: ah each entry only overrides only one flag, not the entire set.
I guess that justifies the stack since it just maintains historical behavior.
It does look like we don't have unit tests that set more than one feature flag yet..

LGTM, queuing

File: Cargo.lock (hunk 1 above)
Comment (danielrainer) [reply Cargo.lock:358]: > use present tense to describe present state ("prior to this commit") and use imperative mood (instead of passive) to describe changes made by this commit.

I generally share these preferences, but in some cases I like being extra explicit about what is old and what is new behavior to avoid confusion. Present tense is also useful for explanations of the new behavior, so sometimes I think it would be unnecessarily ambiguous to also use it for describing old behavior.

> I guess that may not work reliably for tests that spawn new threads, because AIUI, they will not inherit the thread-local override.

Yes, these overrides break if tests spawn threads. I'm not a fan of the thread-local approach, but chose to limit the changes in this PR by not changing this aspect of the code. Global overrides would solve this, but require locking, which would also need to happen for non-test code, so I'm not sure if we want to switch to that. The semantics of one thread overriding features in another one are also tricky and race-condition-prone, so switching from thread-local to global overrides would make the code harder to reason about and write correctly. Using a global lock to prevent multiple tests from running at the same time reduces the race condition problems, but we can still get race conditions from a single test which spawns threads which override features.

One thing we could consider is going with an `OnceLock`-based approach for `FEATURES`, where all code which performs feature tests would have to explicitly initialize `FEATURES` exactly once, so in practice do it early in the binaries and at the start of each test (or at least each tests which needs feature tests). This would allow getting rid of overrides and `AtomicBool`s, but we'd have to lock `FEATURES` for performing tests and address #12611 such that we can properly initialize `FEATURES` before querying. There might be more than the one feature test before `FEATURES` is initialized. I only noticed the one I mentioned because it's the first one, so that's where my experimental code panicked.
